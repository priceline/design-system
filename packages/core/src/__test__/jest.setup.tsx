import '@testing-library/jest-dom'
import 'jest-styled-components'

global.requestAnimationFrame =
  global.requestAnimationFrame ||
  function _raf(cb) {
    return setTimeout(cb, 0)
  }
