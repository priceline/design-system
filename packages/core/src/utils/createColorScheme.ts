import { ColorSchemes, PaletteFamilies } from '../theme'

export const createColorScheme = ({ palette }: { palette: PaletteFamilies }): ColorSchemes => {
  const {
    primary,
    secondary,
    background,
    success,
    warning,
    caution,
    text,
    highlight,
    promoPrimary,
    promoSecondary,
    alert,
  } = palette

  return {
    // Primary
    ///////////////////////////////////////////////
    primary: {
      background: primary.base,
      backgroundName: 'primary.base',
      foreground: background.lightest,
    },
    primaryLight: {
      background: primary.light,
      backgroundName: 'primary.light',
      foreground: primary.base,
    },
    primaryLightest: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: primary.base,
    },
    primaryDark: {
      background: primary.dark,
      backgroundName: 'primary.dark',
      foreground: background.lightest,
    },
    primaryShade: {
      background: primary.shade,
      backgroundName: 'primary.shade',
      foreground: background.lightest,
    },
    primaryDarkOnLight: {
      background: primary.light,
      backgroundName: 'primary.light',
      foreground: primary.dark,
    },

    // Secondary
    ///////////////////////////////////////////////
    secondary: {
      background: secondary.base,
      backgroundName: 'secondary.base',
      foreground: background.lightest,
    },
    secondaryLight: {
      background: secondary.light,
      backgroundName: 'secondary.light',
      foreground: secondary.base,
    },
    secondaryLightest: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: secondary.base,
    },
    secondaryDark: {
      background: secondary.dark,
      backgroundName: 'secondary.dark',
      foreground: background.lightest,
    },
    secondaryDarkOnLight: {
      background: secondary.light,
      backgroundName: 'secondary.light',
      foreground: secondary.dark,
    },

    // Neutral
    ///////////////////////////////////////////////
    neutral: {
      background: background.dark,
      backgroundName: 'background.dark',
      foreground: background.lightest,
    },
    neutralLight: {
      background: background.base,
      backgroundName: 'background.base',
      foreground: background.dark,
    },
    neutralLightest: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: background.dark,
    },
    neutralDark: {
      background: background.darkest,
      backgroundName: 'background.darkest',
      foreground: background.lightest,
    },
    neutralDarkOnLight: {
      background: background.light,
      backgroundName: 'background.light',
      foreground: background.darkest,
    },
    neutralDarkOnLightest: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: background.darkest,
    },

    // Success
    ///////////////////////////////////////////////
    success: {
      background: success.base,
      backgroundName: 'success.base',
      foreground: background.lightest,
    },
    successLight: {
      background: success.light,
      backgroundName: 'success.light',
      foreground: success.base,
    },
    successLightest: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: success.base,
    },
    successDark: {
      background: success.dark,
      backgroundName: 'success.dark',
      foreground: background.lightest,
    },
    successDarkOnLight: {
      background: success.light,
      backgroundName: 'success.light',
      foreground: success.dark,
    },

    // Warning
    ///////////////////////////////////////////////
    warning: {
      background: warning.base,
      backgroundName: 'warning.base',
      foreground: background.lightest,
    },
    warningLight: {
      background: warning.light,
      backgroundName: 'warning.light',
      foreground: warning.base,
    },
    warningLightest: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: warning.base,
    },

    // Caution
    //////////////////////////////////////////////
    cautionLight: {
      background: caution.light,
      backgroundName: 'caution.light',
      foreground: text.base,
    },

    // Highlight
    //////////////////////////////////////////////
    highlightLight: {
      background: highlight.light,
      backgroundName: 'highlight.light',
      foreground: secondary.dark,
    },

    // Promo
    //////////////////////////////////////////////
    promo: {
      background: promoPrimary.base,
      backgroundName: 'promoPrimary.base',
      foreground: text.lightest,
    },
    promoLight: {
      background: promoPrimary.light,
      backgroundName: 'promoPrimary.light',
      foreground: promoPrimary.base,
    },
    promoLightest: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: promoPrimary.base,
    },
    promoDark: {
      background: promoPrimary.dark,
      backgroundName: 'promoPrimary.dark',
      foreground: text.lightest,
    },
    promoDarkOnLight: {
      background: promoPrimary.light,
      backgroundName: 'promoPrimary.light',
      foreground: promoPrimary.dark,
    },

    // Promo Secondary
    //////////////////////////////////////////////
    promoSecondary: {
      background: promoSecondary.base,
      backgroundName: 'promoSecondary.base',
      foreground: text.lightest,
    },
    promoSecondaryDark: {
      background: promoSecondary.dark,
      backgroundName: 'promoSecondary.dark',
      foreground: text.lightest,
    },

    // Alert
    //////////////////////////////////////////////
    alert: {
      background: alert.base,
      backgroundName: 'alert.base',
      foreground: background.lightest,
    },
    alertLight: {
      background: alert.light,
      backgroundName: 'alert.light',
      foreground: alert.base,
    },
    alertLightest: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: alert.base,
    },
    alertDarkOnLight: {
      background: alert.light,
      backgroundName: 'alert.light',
      foreground: alert.tone,
    },

    // Heading
    //////////////////////////////////////////////
    heading: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: text.heading,
    },
    headingBase: {
      background: background.base,
      backgroundName: 'background.base',
      foreground: text.heading,
    },
    headingLight: {
      background: background.light,
      backgroundName: 'background.light',
      foreground: text.heading,
    },
    headingOnPrimaryLight: {
      background: primary.light,
      backgroundName: 'primary.light',
      foreground: text.heading,
    },
  }
}
