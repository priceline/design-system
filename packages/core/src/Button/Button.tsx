import { themeGet } from '@styled-system/theme-get'
import React, { type ComponentPropsWithRef } from 'react'
import styled, { css } from 'styled-components'
import {
  BoxShadowProps,
  HeightProps,
  type ResponsiveValue,
  SpaceProps,
  WidthProps,
  borderRadius,
  boxShadow,
  compose,
  fontSize,
  height,
  space,
  width,
} from 'styled-system'
import { Flex, type FlexProps } from '../Flex/Flex'
import { type BorderRadius, type BoxShadowSize } from '../theme'
import { boxShadowAttrs } from '../utils/attrs/boxShadowAttrs'
import { applySizes, applyVariations, borders, getPaletteColor, getTextColorOn } from '../utils/utils'

/**
 * @public
 */
export const borderRadiusButtonValues = ['none', 'sm', 'md', 'lg', 'xl', '2xl', '3xl', 'full']

const isValidBorderRadius = (size) => size && borderRadiusButtonValues.includes(size)

const sizes = {
  small: css`
    border-radius: ${(props) =>
      themeGet(`borderRadii.${isValidBorderRadius(props.borderRadius) ? props.borderRadius : 'action-sm'}`)(
        props
      )};
    font-size: ${themeGet('fontSizes.0')}px;
    padding: 7px 12px;
  `,
  medium: css`
    border-radius: ${(props) =>
      themeGet(`borderRadii.${isValidBorderRadius(props.borderRadius) ? props.borderRadius : 'action-md'}`)(
        props
      )};
    font-size: ${themeGet('fontSizes.1')}px;
    padding: 9.5px 18px;
  `,
  large: css`
    border-radius: ${(props) =>
      themeGet(`borderRadii.${isValidBorderRadius(props.borderRadius) ? props.borderRadius : 'action-lg'}`)(
        props
      )};
    font-size: ${themeGet('fontSizes.2')}px;
    padding: 12px 22px;
  `,
  extraLarge: css`
    border-radius: ${(props) =>
      themeGet(`borderRadii.${isValidBorderRadius(props.borderRadius) ? props.borderRadius : 'action-xl'}`)(
        props
      )};
    font-size: ${themeGet('fontSizes.2')}px;
    padding: 16px 22px;
  `,
}

const variations = {
  fill: css`
    background-color: ${(props) => getPaletteColor(props.disabled ? 'light' : 'base')(props)};
    color: ${(props) => getTextColorOn(props.disabled ? 'light' : 'base')(props)};

    &:hover {
      background-color: ${(props) => getPaletteColor(props.disabled ? 'light' : 'dark')(props)};
      ${(props) => (props.disabled ? '' : `color: ${getTextColorOn('dark')(props)};`)}
    }
    &:focus-visible {
      outline: ${(props) => `0px solid ${getPaletteColor(props.disabled ? '' : 'dark')(props)}`};
      box-shadow: ${(props) => ` 0 0 0 2px  ${getPaletteColor(props.disabled ? '' : 'dark')(props)}`};
    }
  `,
  link: css`
    -webkit-font-smoothing: inherit;
    font-weight: ${(props) => props.theme.fontWeights.medium};
    color: ${getPaletteColor('base')};
    line-height: ${(props) => props.theme.lineHeights.standard};
    vertical-align: inherit;
    padding: 0;
    background-color: transparent;
    &:hover {
      color: ${getPaletteColor('dark')};
      text-decoration: underline;
    }
    &&&:disabled {
      background-color: transparent;
    }
  `,
  outline: css`
    color: ${(props) => getPaletteColor(props.disabled ? 'light' : 'base')(props)};
    box-shadow: inset 0 0 0 2px ${(props) => getPaletteColor(props.disabled ? 'light' : 'base')(props)};
    background-color: transparent;

    &:hover {
      background-color: transparent;
      ${(props) =>
        props.disabled
          ? ''
          : `
      color: ${getPaletteColor('dark')(props)};
      box-shadow: inset 0 0 0 2px ${getPaletteColor('dark')(props)};
    `}
    }
  `,
  plain: css`
    background-color: transparent;
    color: ${getPaletteColor('base')};
    &:hover {
      background-color: ${getPaletteColor('background.base')};
      color: ${getPaletteColor('dark')};
    }
    &:focus-visible {
      outline: 0px solid ${getPaletteColor('dark')};
      box-shadow: 0 0 0 2px ${getPaletteColor('dark')};
    }
  `,
  subtle: css`
    background-color: ${getPaletteColor('background.base')};
    color: ${getPaletteColor('base')};
    &:hover {
      background-color: ${getPaletteColor('background.tint')};
      color: ${getPaletteColor('dark')};
    }
    &:focus-visible {
      outline: 0px solid ${getPaletteColor('dark')};
      box-shadow: 0 0 0 2px ${getPaletteColor('dark')};
    }
  `,
  white: css`
    background-color: ${getPaletteColor('background.lightest')};
    color: ${getPaletteColor('base')};
    &:hover {
      background-color: ${getPaletteColor('background.base')};
      color: ${getPaletteColor('dark')};
    }
    &:focus-visible {
      outline: 0px solid ${getPaletteColor('dark')};
      box-shadow: 0 0 0 2px ${getPaletteColor('dark')};
    }
  `,
  lightFill: css`
    background-color: ${getPaletteColor('light')};
    color: ${getPaletteColor('base')};
    ${(props) =>
      props.disabled
        ? ''
        : `
      &:hover {
        background-color: ${getPaletteColor('light')(props)};
        color: ${getPaletteColor('dark')(props)};
      }`}
    &:focus-visible {
      outline: ${(props) => `0px solid ${getPaletteColor(props.disabled ? '' : 'dark')(props)}`};
      box-shadow: ${(props) => ` 0 0 0 2px  ${getPaletteColor(props.disabled ? '' : 'dark')(props)}`};
    }
  `,
  input: css`
    appearance: none;
    background-color: transparent;
    border-style: solid;
    border-width: 1px;
    color: ${getPaletteColor('text.base')};
    display: block;
    font-family: inherit;
    font-weight: normal;
    line-height: normal;
    margin: 0;
    padding: 14px 12px;
    text-align: left;
    width: 100%;

    &:disabled {
      background-color: ${getPaletteColor('background.light')};
      color: ${getPaletteColor('text.light')};
    }

    ${(props) => borders({ ...props, color: undefined })}
    ${(props) => compose(space, fontSize, borderRadius)(props)}
  `,
}

/**
 * @public
 */
export type ButtonSizes = 'small' | 'medium' | 'large' | 'extraLarge'

/**
 * @public
 */
export type ButtonVariations =
  | 'fill'
  | 'link'
  | 'outline'
  | 'plain'
  | 'subtle'
  | 'white'
  | 'lightFill'
  | 'input'

/**
 * @public
 */
export type StyledButtonProps = ButtonProps & { hasChildren: boolean }

/**
 * @public
 */
export type ButtonProps = WidthProps &
  HeightProps &
  SpaceProps &
  BoxShadowProps &
  ComponentPropsWithRef<'button'> & {
    color?: string
    variation?: ButtonVariations
    size?: ButtonSizes | ButtonSizes[]
    borderRadius?: BorderRadius
    boxShadowSize?: ResponsiveValue<BoxShadowSize>
    autoFocus?: boolean
    IconLeft?: React.Component | React.FC
    IconRight?: React.Component | React.FC
    flexProps?: FlexProps
  }

export const buttonStyles = css`
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  vertical-align: middle;
  text-align: center;
  text-decoration: none;
  font-family: inherit;
  font-weight: ${(props) => props.theme.fontWeights.bold};
  line-height: 1.5;
  cursor: ${(props) => (props.disabled ? 'default' : 'pointer')};
  border-radius: ${(props) =>
    themeGet(`borderRadii.${isValidBorderRadius(props.borderRadius) ? props.borderRadius : 'action-md'}`)(
      props
    )};
  border-width: 0;
  border-style: solid;

  &:disabled {
    cursor: not-allowed;
    color: ${getPaletteColor('background.tone')};
    background-color: ${getPaletteColor('background.base')};
  }

  ${({ theme }) => applySizes(sizes, 'medium', theme.mediaQueries)};
  ${applyVariations('Button', variations)};
`

const iconButtonPaddings = {
  IconLeft: {
    small: { p: '6px 14px 6px 10px' },
    medium: { p: '8px 20px 8px 16px' },
    large: { p: '12px 24px 12px 20px' },
    extraLarge: { p: '16px 24px 16px 20px' },
  },
  IconRight: {
    small: { p: '6px 10px 6px 14px' },
    medium: { p: '8px 16px 8px 20px' },
    large: { p: '12px 20px 12px 24px' },
    extraLarge: { p: '16px 20px 16px 24px' },
  },
  both: {
    small: { p: '6px 10px' },
    medium: { p: '8px 16px' },
    large: { p: '12px 20px' },
    extraLarge: { p: '16px 20px' },
  },
}
const iconOnlyButtonPaddings = {
  small: { px: '6px', py: '6px' },
  medium: { px: '8px', py: '8px' },
  large: { px: '12px', py: '12px' },
  extraLarge: { px: '16px', py: '16px' },
}
const iconButtonSizes = {
  small: 20,
  medium: 24,
  large: 24,
  extraLarge: 24,
}

// eslint-disable-next-line @typescript-eslint/naming-convention
function getPaddingProps({ IconLeft, IconRight, size, variation, hasChildren }: StyledButtonProps) {
  const hasIcon = !!(IconLeft || IconRight)
  const sizeIndex = Array.isArray(size) ? size[0] : size

  if (variation === 'link') {
    return { p: 0 }
  }

  if (hasIcon) {
    if (hasChildren) {
      if (IconLeft && IconRight) {
        return iconButtonPaddings.both[sizeIndex]
      } else if (IconLeft) {
        return iconButtonPaddings.IconLeft[sizeIndex]
      } else if (IconRight) {
        return iconButtonPaddings.IconRight[sizeIndex]
      }
      return {}
    } else {
      return iconOnlyButtonPaddings?.[sizeIndex] ?? {}
    }
  }

  return {}
}

/**
 * Use the <Button /> component to render a primitive button. Use the `variation` prop to change the look of the button.
 */
const StyledButton: React.FC<StyledButtonProps> = styled.button.attrs((props) => {
  const {
    width,
    height,
    title,
    'aria-label': ariaLabel,
    borderRadius,
    IconLeft,
    IconRight,
    size,
    variation,
    hasChildren,
  } = props

  const paddingProps = getPaddingProps({ IconLeft, IconRight, size, variation, hasChildren })

  return {
    borderRadius,
    ...boxShadowAttrs(props),
    width,
    height,
    'aria-label': ariaLabel || title,
    ...paddingProps,
  }
})`
  ${buttonStyles}

  ${(props) => compose(width, height, space, boxShadow)(props)}
`

// eslint-disable-next-line @typescript-eslint/naming-convention
const ButtonIcon = ({ Component, ...props }) => {
  return Component ? <Component {...props} /> : null
}

/**
 * @public
 */
export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      color = 'primary',
      children,
      IconLeft,
      IconRight,
      size = 'medium',
      flexProps = {},
      variation = 'fill',
      ...restProps
    },
    ref
  ) => {
    const hasChildren = React.Children.toArray(children).length > 0
    const sizeIndex = Array.isArray(size) ? size[0] : size
    const iconSize = iconButtonSizes?.[sizeIndex] ?? 24

    return (
      <StyledButton
        {...restProps}
        color={color}
        variation={variation}
        hasChildren={hasChildren}
        IconLeft={IconLeft}
        IconRight={IconRight}
        ref={ref}
        flexProps={flexProps}
        size={size}
      >
        <Flex alignItems='center' justifyContent='center' {...flexProps}>
          <ButtonIcon Component={IconLeft} size={iconSize} mr={children ? 2 : 0} />
          {children}
          <ButtonIcon Component={IconRight} size={iconSize} ml={children ? 2 : 0} />
        </Flex>
      </StyledButton>
    )
  }
)

Button.displayName = 'Button'
