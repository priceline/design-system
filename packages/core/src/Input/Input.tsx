import React, { type ComponentPropsWithRef, ForwardRefExoticComponent, RefAttributes } from 'react'
import styled, { css } from 'styled-components'
import {
  FontSizeProps,
  SpaceProps,
  ZIndexProps,
  borderRadius,
  compose,
  fontSize,
  space,
  zIndex,
} from 'styled-system'
import { Text, type TextProps } from '../Text/Text'
import { borderRadiusAttrs } from '../utils/attrs/borderRadiusAttrs'
import { applySizes, applyVariations, borders, getPaletteColor } from '../utils/utils'

const sizes = {
  sm: css`
    padding: 6px 12px 7px 12px;
  `,
  md: css`
    padding: 10px 12px 11px 12px;
  `,
  lg: css`
    padding: 14px 12px 14px 12px;
  `,
  xl: css`
    padding: 18px 12px 19px 12px;
  `,
}

const StyledInput = styled.input.attrs(borderRadiusAttrs)`
  position: relative;
  appearance: none;
  display: block;
  width: 100%;
  font-family: inherit;
  color: ${getPaletteColor('text.base')};
  background-color: transparent;
  border-width: 1px;
  border-style: solid;
  margin: 0;

  ::placeholder {
    color: ${getPaletteColor('text.light')};
  }

  ::-ms-clear {
    display: none;
  }

  &:disabled {
    background-color: ${getPaletteColor('background.light')};
    color: ${getPaletteColor('text.light')};
    cursor: not-allowed;
  }

  ${({ theme }) => applySizes(sizes, undefined, theme.mediaQueries)};
  ${applyVariations('Input')}
  ${borders}

  ${(props) => compose(space, fontSize, borderRadius, zIndex)(props)}
`

const INPUT_ERROR_TEXT = 'InputHelperText'

/**
 * @public
 */
export type InputHelperTextProps = TextProps & {
  children?: React.ReactNode
}

/**
 * @public
 */
export type InputProps = SpaceProps &
  FontSizeProps &
  ZIndexProps &
  Omit<ComponentPropsWithRef<'input'>, 'size'> & {
    children?: React.ReactNode
    helperText?: React.ReactElement<
      InputHelperTextProps,
      string | React.JSXElementConstructor<InputHelperTextProps>
    >
    color?: string
    size?: 'sm' | 'md' | 'lg' | 'xl'
    borderRadius?: string
  }

/**
 * @public
 */
export type InputWithRef = ForwardRefExoticComponent<Omit<InputProps, 'ref'> & RefAttributes<unknown>> & {
  isField: boolean
  HelperText: React.FC<InputHelperTextProps>
}

/**
 * @public
 */
export const Input = React.forwardRef<HTMLInputElement, InputProps>((props, ref) => {
  const { helperText, color, ...restProps } = props
  return (
    <>
      <StyledInput {...restProps} color={color} ref={ref} />
      {helperText &&
        React.cloneElement(helperText, {
          color: helperText.props.color || color,
        })}
    </>
  )
}) as InputWithRef

const HelperText: React.FC<InputHelperTextProps> = styled(Text).attrs(() => ({
  mt: 2,
  fontSize: 1,
}))``

Input.HelperText = (props: InputHelperTextProps) => <HelperText {...props}>{props.children}</HelperText>

Input.HelperText.displayName = INPUT_ERROR_TEXT

Input.displayName = 'Input'
Input.isField = true
Input.defaultProps = {
  fontSize: [2, null, 1],
  borderRadius: 'lg',
  size: 'lg',
}
