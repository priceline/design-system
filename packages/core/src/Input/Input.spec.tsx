import { fireEvent, render, screen } from '../__test__/testing-library'
import { Button } from '../Button/Button'
import ForwardRefDemo from '../storybook/utils/ForwardRefsDemo'
import { theme } from '../theme/theme'
import { createTheme } from '../utils/createTheme'
import { getPaletteColor } from '../utils/utils'
import { Input } from './Input'

const id = 'fake-test-id'

describe('Input', () => {
  test('it renders', () => {
    const { asFragment } = render(<Input id={id} />)
    expect(asFragment()).toMatchSnapshot()
  })
  test('it renders an input element with a red border with a color prop is passed', () => {
    const { asFragment } = render(<Input id={id} color='red' />)
    expect(asFragment()).toMatchSnapshot()
  })
  test('it renders an input element with a really large padding and margin', () => {
    const { asFragment } = render(<Input id={id} p={4} m={4} />)
    expect(asFragment()).toMatchSnapshot()
  })
  test('it renders an input element with large text', () => {
    const { asFragment } = render(<Input id={id} fontSize={4} />)
    expect(asFragment()).toMatchSnapshot()
  })
  test('forwards ref to the input', () => {
    const target = 'TARGET'
    const placeholder = 'Apples'
    function refChild(dsRef) {
      function onClick() {
        dsRef.current.focus()
      }
      return (
        <>
          <Input id={id} ref={dsRef} placeholder={placeholder} />
          <Button onClick={onClick} mt={3}>
            {target}
          </Button>
        </>
      )
    }
    const { getByText, getByPlaceholderText } = render(<ForwardRefDemo refChild={refChild} />)
    fireEvent.click(getByText(target))
    expect(getByPlaceholderText(placeholder)).toHaveFocus()
  })

  test('it renders disabled', () => {
    const paletteTheme = createTheme(theme)
    render(<Input id={id} disabled placeholder='Disabled Input' />)

    const input = screen.getByPlaceholderText('Disabled Input')
    expect(input).toBeDisabled()
    expect(input).toHaveStyleRule(
      'background-color',
      getPaletteColor('background.light')({ theme: paletteTheme }),
      { modifier: ':disabled' }
    )
    expect(input).toHaveStyleRule('color', getPaletteColor('text.light')({ theme: paletteTheme }), {
      modifier: ':disabled',
    })
    expect(input).toHaveStyleRule('cursor', 'not-allowed', { modifier: ':disabled' })
  })

  describe('helper text', function () {
    test('renders with same color as Input', () => {
      const helperText = 'hello world'
      const color = 'error.base'
      const paletteTheme = createTheme(theme)
      const { getByText } = render(
        <Input id={id} color={color} helperText={<Input.HelperText>{helperText}</Input.HelperText>} />
      )
      const helperTextNode = getByText(helperText)
      expect(helperTextNode).not.toBeNull()
      expect(helperTextNode).toHaveStyleRule('color', getPaletteColor(color)({ theme: paletteTheme }))
    })

    test('can override the color from Input', () => {
      const helperText = 'hello world'
      const placeholder = 'click me'
      const color = 'error.base'
      const helperTextColor = 'secondary.base'
      const paletteTheme = createTheme(theme)
      const { getByText, getByPlaceholderText } = render(
        <Input
          id={id}
          color={color}
          placeholder={placeholder}
          helperText={<Input.HelperText color={helperTextColor}>{helperText}</Input.HelperText>}
        />
      )
      const helperTextNode = getByText(helperText)
      expect(helperTextNode).not.toBeNull()
      expect(helperTextNode).toHaveStyleRule(
        'color',
        getPaletteColor(helperTextColor)({ theme: paletteTheme })
      )
      expect(getByPlaceholderText(placeholder)).toHaveStyleRule(
        'border-color',
        getPaletteColor(color)({ theme: paletteTheme })
      )
    })
  })
})
