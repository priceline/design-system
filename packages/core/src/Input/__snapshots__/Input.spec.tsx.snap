// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input it renders 1`] = `
.c0 {
  position: relative;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: block;
  width: 100%;
  font-family: inherit;
  color: #001833;
  background-color: transparent;
  border-width: 1px;
  border-style: solid;
  margin: 0;
  padding: 14px 12px 14px 12px;
  border-color: #c0cad5;
  font-size: 16px;
  border-radius: 12px;
}

.c0::-webkit-input-placeholder {
  color: #4f6f8f;
}

.c0::-moz-placeholder {
  color: #4f6f8f;
}

.c0:-ms-input-placeholder {
  color: #4f6f8f;
}

.c0::placeholder {
  color: #4f6f8f;
}

.c0::-ms-clear {
  display: none;
}

.c0:disabled {
  background-color: #f4f6f8;
  color: #4f6f8f;
  cursor: not-allowed;
}

.c0:focus {
  outline: 0;
  border-color: #0068ef;
  box-shadow: 0 0 0 2px #0068ef;
}

@media screen and (min-width:32em) {

}

@media screen and (min-width:40em) {
  .c0 {
    font-size: 14px;
  }
}

<input
  className="c0"
  fontSize={
    Array [
      2,
      null,
      1,
    ]
  }
  id="fake-test-id"
  size="lg"
/>
`;

exports[`Input it renders an input element with a really large padding and margin 1`] = `
.c0 {
  position: relative;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: block;
  width: 100%;
  font-family: inherit;
  color: #001833;
  background-color: transparent;
  border-width: 1px;
  border-style: solid;
  margin: 0;
  padding: 14px 12px 14px 12px;
  border-color: #c0cad5;
  padding: 32px;
  margin: 32px;
  font-size: 16px;
  border-radius: 12px;
}

.c0::-webkit-input-placeholder {
  color: #4f6f8f;
}

.c0::-moz-placeholder {
  color: #4f6f8f;
}

.c0:-ms-input-placeholder {
  color: #4f6f8f;
}

.c0::placeholder {
  color: #4f6f8f;
}

.c0::-ms-clear {
  display: none;
}

.c0:disabled {
  background-color: #f4f6f8;
  color: #4f6f8f;
  cursor: not-allowed;
}

.c0:focus {
  outline: 0;
  border-color: #0068ef;
  box-shadow: 0 0 0 2px #0068ef;
}

@media screen and (min-width:32em) {

}

@media screen and (min-width:40em) {
  .c0 {
    font-size: 14px;
  }
}

<input
  className="c0"
  fontSize={
    Array [
      2,
      null,
      1,
    ]
  }
  id="fake-test-id"
  size="lg"
/>
`;

exports[`Input it renders an input element with a red border with a color prop is passed 1`] = `
.c0 {
  position: relative;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: block;
  width: 100%;
  font-family: inherit;
  color: #001833;
  background-color: transparent;
  border-width: 1px;
  border-style: solid;
  margin: 0;
  padding: 14px 12px 14px 12px;
  border-color: #c00;
  font-size: 16px;
  border-radius: 12px;
}

.c0::-webkit-input-placeholder {
  color: #4f6f8f;
}

.c0::-moz-placeholder {
  color: #4f6f8f;
}

.c0:-ms-input-placeholder {
  color: #4f6f8f;
}

.c0::placeholder {
  color: #4f6f8f;
}

.c0::-ms-clear {
  display: none;
}

.c0:disabled {
  background-color: #f4f6f8;
  color: #4f6f8f;
  cursor: not-allowed;
}

.c0:focus {
  outline: 0;
  border-color: #c00;
  box-shadow: 0 0 0 2px #c00;
}

@media screen and (min-width:32em) {

}

@media screen and (min-width:40em) {
  .c0 {
    font-size: 14px;
  }
}

<input
  className="c0"
  color="red"
  fontSize={
    Array [
      2,
      null,
      1,
    ]
  }
  id="fake-test-id"
  size="lg"
/>
`;

exports[`Input it renders an input element with large text 1`] = `
.c0 {
  position: relative;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: block;
  width: 100%;
  font-family: inherit;
  color: #001833;
  background-color: transparent;
  border-width: 1px;
  border-style: solid;
  margin: 0;
  padding: 14px 12px 14px 12px;
  border-color: #c0cad5;
  font-size: 24px;
  border-radius: 12px;
}

.c0::-webkit-input-placeholder {
  color: #4f6f8f;
}

.c0::-moz-placeholder {
  color: #4f6f8f;
}

.c0:-ms-input-placeholder {
  color: #4f6f8f;
}

.c0::placeholder {
  color: #4f6f8f;
}

.c0::-ms-clear {
  display: none;
}

.c0:disabled {
  background-color: #f4f6f8;
  color: #4f6f8f;
  cursor: not-allowed;
}

.c0:focus {
  outline: 0;
  border-color: #0068ef;
  box-shadow: 0 0 0 2px #0068ef;
}

<input
  className="c0"
  fontSize={4}
  id="fake-test-id"
  size="lg"
/>
`;
